File: 002/Erlang/problem002.erl
-module(problem002). 
-export([fibSum/1, takeFibs/1]).

fibSum(N) -> lists:sum(lists:filter(fun(E) -> (E rem 2) =:= 0 end, takeFibs(N))).

takeFibs(N) ->
	if
	  N == 0 -> [];
	  true   -> takeFibs(N, 2, 1, [])
    end.

takeFibs(N, First, Second, Fibs) ->
	if
		First > N -> [Second | Fibs];
		true      -> takeFibs(N, First + Second, First, [Second | Fibs])
	end.

File: 002/Prolog/problem002.pl
fibSum(N, Sum) :-
	takeFibs(N, Fibs),
	include(even, Fibs, EvenFibs),
	sumlist(EvenFibs, Sum).

even(N) :- N mod 2 =:= 0.

takeFibs(N, L) :-
	N =:= 0 -> L = [];
	N > 0   -> takeFibs(N, 2, 1, [], L).
	
takeFibs(N, First, Second, Incoming, Outgoing) :-
	First > N  -> Outgoing = [Second | Incoming];
	First =< N -> Next is First+Second, takeFibs(N, Next, First, [Second | Incoming], Outgoing).
